packages/mom5/package.py

```diff
@@ -0,0 +1,317 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# Copyright 2023 ACCESS-NRI
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import install, join_path, mkdirp
+
+# https://spack.readthedocs.io/en/latest/build_systems/makefilepackage.html
+class Mom5(MakefilePackage):
+    """MOM is a numerical ocean model based on the hydrostatic primitive equations."""
+
+    homepage = "https://www.access-nri.org.au"
+    git = "https://github.com/ACCESS-NRI/mom5.git"
+
+    maintainers = ["harshula"]
+
+    version("spack-build", branch="spack-build")
+
+    # Depend on virtual package "mpi".
+    depends_on("mpi")
+    depends_on("oasis3-mct")
+    depends_on("datetime-fortran")
+    depends_on("netcdf-fortran@4.5.2:")
+    depends_on("netcdf-c@4.7.4:")
+    # TODO: For initial verification we are going to use static pio.
+    #       Eventually we plan to move to shared pio
+    # ~shared requires: https://github.com/spack/spack/pull/34837
+    depends_on("parallelio~pnetcdf~timing~shared")
+    depends_on("libaccessom2")
+    depends_on("hdf5")
+
+    phases = ["edit", "build", "install"]
+
+    def edit(self, spec, prefix):
+
+        srcdir = self.stage.source_path
+        makeinc_path = join_path(srcdir, "bin", "mkmf.template.spack")
+        config = {}
+
+        incs = (spec["oasis3-mct"].headers).cpp_flags + "/psmile.MPI1" + " "
+        for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]:
+            incs += (spec[l].headers).cpp_flags + " "
```

Comment:
```suggestion
        incs = [spec[l].headers).cpp_flags for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]]
        incs = " ".join(incs)
```
Using a list comprehension is "pythonic" as is using a join to convert a list to a space separated string.

Reply:
Does your suggestion traverse the list twice? Remember, we want to take a list and create a string in a single traversal of the list. List comprehension is great when the input and the desired output are lists.
Also, the code is more comprehensible the way it is.

Reply:
Yes I am suggesting traversing the list twice.

I think it is clearer to use two steps. It is just an opinion. If you disagree feel free to ignore this suggestion.

Reply:
Note: to be specific the second traversal isn't the same list, the `psmile.MPI1` include at the beginning and the various `cpp_flags` lookups have taken place for the listed dependencies. And yes, I am suggesting traversing **that** list.

Reply:
```
>>> import timeit
>>> h = """\
... lut = {"parallelio": "a", "oasis3-mct": "b", "libaccessom2": "c", "netcdf-fortran": "d"}
... incs = ""
... for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]:
...     incs += lut[l] + " "
... """
>>> a = """\
... lut = {"parallelio": "a", "oasis3-mct": "b", "libaccessom2": "c", "netcdf-fortran": "d"}
... incs = [lut[l] for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]]
... inc = " ".join(incs)
... """
>>> timeit.timeit(stmt=h, number=10000000)
2.131908037001267
>>> timeit.timeit(stmt=a, number=10000000)
2.251114501967095
>>> timeit.timeit(stmt=h, number=10000000)
2.135053211997729
>>> timeit.timeit(stmt=a, number=10000000)
2.2056077060406096
>>> timeit.timeit(stmt=h, number=10000000)
2.1651839180267416
>>> timeit.timeit(stmt=a, number=10000000)
2.2446520270314068
```

Reply:
I don't think performance is as important as clarity. I don't necessarily think the list comprehension is much clearer than the loop, simply that once you're just doing a lookup and there is no other logic it might be easier to use a comprehension, but I think this is perfectly fine:
```python
lut = {"parallelio": "a", "oasis3-mct": "b", "libaccessom2": "c", "netcdf-fortran": "d"}
incs = []
for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]:
     incs += lut[l]
inc = " ".join(incs)
```

Agreed there are other methods: map, formatted string, class with a method
```python
>>> " ".join(map(lambda l: lut[l], ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]))
'a b c d'
```
but I don't think `map` is necessarily clearer.

Reply:
This avoids appending a space, avoids traversing the list twice and uses join():
```
        istr = join_path((spec["oasis3-mct"].headers).cpp_flags, "psmile.MPI1")
        ideps = ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]
        incs = " ".join([istr] + [(spec[d].headers).cpp_flags for d in ideps])
```

Reply:
Indeed. Looks clean and clear.